tmt/steps/provision/__init__.py

```diff
@@ -2507,77 +2564,96 @@ def push(
             self._run_guest_command(
-                Command(
-                    *cmd,
-                    *options,
-                    "-e",
-                    self._ssh_command.to_element(),
-                    source,
-                    f"{self._ssh_guest}:{destination}",
-                ),
+                Command(*full_cmd_list),
                 silent=True,
             )
 
-        # Try to push twice, check for rsync after the first failure
         try:
-            rsync()
-        except tmt.utils.RunError:
+            _execute_rsync_operation()
+            self.debug(
+                f"Successfully pushed '{source_str}' to '{self._ssh_guest}:{dest_str}'", level=2
+            )
+        except tmt.utils.RunError as first_error:
+            self.debug(f"Initial rsync attempt failed: {first_error}", level=2)
+            if self.is_dry_run:
+                self.warn("Rsync failed during dry run, not retrying.")
+                raise
+
+            self.debug("Checking rsync on guest and attempting retry...")
             try:
-                if self._check_rsync() == CheckRsyncOutcome.ALREADY_INSTALLED:
-                    raise
-                rsync()
-            except tmt.utils.RunError:
-                # Provide a reasonable error to the user
+                rsync_check_outcome = self._check_rsync()
+
+                if rsync_check_outcome == CheckRsyncOutcome.ALREADY_INSTALLED:
+                    self.debug(
+                        "rsync reported as already installed. The original error is likely "
+                        "due to connection, permissions, or rsync arguments."
+                    )
+                    raise first_error
+
+                self.debug("rsync check indicates a retry is warranted. Retrying push operation.")
+                _execute_rsync_operation()
+                self.debug(
+                    f"Pushed '{source_str}' to '{self._ssh_guest}:{dest_str}' on retry.",
+                    level=2,
+                )
+
+            except tmt.utils.RunError as second_error:
+                error_message = (
+                    f"Failed to push '{source_str}' to guest '{self._ssh_guest}' after retry. "
+                    f"This often indicates issues with login as '{self.user}', "
+                    f"rsync setup on the guest, or permissions.\n"
+                    f"Initial rsync error: {first_error}\n"
+                    f"Subsequent error during check/retry: {second_error}"
+                )
+                self.fail(error_message)
+                raise second_error from first_error
+            except Exception as check_retry_exception:
                 self.fail(
-                    f"Failed to push workdir to the guest. This usually means "
-                    f"that login as '{self.user}' to the guest does not work."
+                    f"An unexpected error during rsync check or retry: {check_retry_exception}"
```

Comment:
No need for including `check_retry_exception` in the error message, the exception is reraised on the next line, it will be recorded.